packages/client-react-streaming/src/registerApolloClient.tsx

```diff
@@ -1,8 +1,24 @@
-import type { ApolloClient } from "@apollo/client/index.js";
+import { ApolloClient } from "@apollo/client/index.js";
 import { cache } from "react";
 
+const seenWrappers = WeakSet
+  ? new WeakSet<{ client: ApolloClient<any> }>()
+  : undefined;
+const seenClients = WeakSet ? new WeakSet<ApolloClient<any>>() : undefined;
+
 export function registerApolloClient(makeClient: () => ApolloClient<any>) {
-  function wrappedMakeClient() {
+  // this function will always create a new object
+  // so we can warn if `cachedMakeWrappedClient` creates a new "wrapper",
+  // but with a `client` property that we have already seen before.
+  // In that case, not every call to `makeClient` would create a new
+  // `ApolloClient` instance.
+  function makeWrappedClient() {
+    return { client: makeClient() };
+  }
```

Comment:
I think I got a little confused at first on why you'd need the wrapper object here instead of just returning `makeClient` directly. Looking at the [React documentation](https://react.dev/reference/react/cache#noun-labs-1201738-(2)), I see this point:

> React will invalidate the cache for all memoized functions for each server request.

I'm assuming this is the reason you're returning a wrapping object rather than the client directly? Is it just harder to detect unique references to `client` otherwise?

If so, perhaps you can update the comment to provide a little more clarity. Something like:

```suggestion
  // React invalidates the cache on each server request, so the wrapping
  // object is needed to properly detect whether the client is a unique 
  // reference or not. We can warn if `cachedMakeWrappedClient` creates a new "wrapper",
  // but with a `client` property that we have already seen before.
  // In that case, not every call to `makeClient` would create a new
  // `ApolloClient` instance.
  function makeWrappedClient() {
    return { client: makeClient() };
  }
```

Do let me know if I'm thinking about this in the wrong way.

Reply:
The user could (and should!) call the memoized `getClient` function a dozen times while working on a request.

So it is normal that this memoized function runs multiple times and returns a `client` instance we have seen before. We need one piece of additional information: "Was this the first time for this request we called `getClient`?" Because only in this case, returning a "known" client instance would be problematic.

So we do something where we **know** that it will be created once per request uniquely - we memoize creating an object (the wrapper).

If we see a new object and a new client - all good.
If we see a known object and a known client - all good.
If we see a new object, but an old client - this is bad, a user is not creating a new client in their `makeClient` function.

Reply:
=> Yeah, I think your comment makes a ton of sense - taking it!